google-cloud-firestore/src/main/java/com/google/cloud/firestore/RateLimiter.java

```diff
@@ -0,0 +1,139 @@
+   * @param startTimeMillis The starting time in epoch milliseconds that the rate limit is based on.
+   *     Used for testing the limiter.
+   */
+  RateLimiter(int initialCapacity, double multiplier, int multiplierMillis, long startTimeMillis) {
+    this.initialCapacity = initialCapacity;
+    this.multiplier = multiplier;
+    this.multiplierMillis = multiplierMillis;
+    this.startTimeMillis = startTimeMillis;
+
+    this.availableTokens = initialCapacity;
+    this.lastRefillTimeMillis = startTimeMillis;
+  }
+
+  public boolean tryMakeRequest(int numOperations) {
+    return tryMakeRequest(numOperations, new Date(0).getTime());
+  }
+
+  /**
+   * Tries to make the number of operations. Returns true if the request succeeded and false
+   * otherwise.
+   *
+   * @param requestTimeMillis The time used to calculate the number of available tokens. Used for
+   *     testing the limiter.
+   */
+  public boolean tryMakeRequest(int numOperations, long requestTimeMillis) {
+    refillTokens(requestTimeMillis);
+    if (numOperations <= availableTokens) {
+      availableTokens -= numOperations;
+      return true;
+    }
+    return false;
+  }
+
+  public long getNextRequestDelayMs(int numOperations) {
+    return getNextRequestDelayMs(numOperations, new Date().getTime());
+  }
+
+  /**
+   * Returns the number of ms needed to make a request with the provided number of operations.
+   * Returns 0 if the request can be made with the existing capacity. Returns -1 if the request is
+   * not possible with the current capacity.
+   *
+   * @param requestTimeMillis The time used to calculate the number of available tokens. Used for
+   *     testing the limiter.
+   */
+  public long getNextRequestDelayMs(int numOperations, long requestTimeMillis) {
+    if (numOperations < availableTokens) {
+      return 0;
+    }
+
+    int capacity = calculateCapacity(requestTimeMillis);
+    if (capacity < numOperations) {
+      return -1;
+    }
+
+    int requiredTokens = numOperations - availableTokens;
+    return (long) Math.ceil((double) (requiredTokens * 1000) / capacity);
+  }
+
+  /**
+   * Refills the number of available tokens based on how much time has elapsed since the last time
+   * the tokens were refilled.
+   *
+   * @param requestTimeMillis The time used to calculate the number of available tokens. Used for
+   *     testing the limiter.
+   */
+  private void refillTokens(long requestTimeMillis) {
+    if (requestTimeMillis >= lastRefillTimeMillis) {
+      long elapsedTime = requestTimeMillis - lastRefillTimeMillis;
+      int capacity = calculateCapacity(requestTimeMillis);
+      int tokensToAdd = (int) Math.floor((elapsedTime * capacity) / 1000);
+      if (tokensToAdd > 0) {
+        availableTokens = Math.min(capacity, availableTokens + tokensToAdd);
+        lastRefillTimeMillis = requestTimeMillis;
+      }
+    } else {
+      throw new IllegalArgumentException(
```

Comment:
Optional: You Preconditions.checkArgument as first statement of this method.

Reply:
Is there any benefit to using `Preconditions.checkArgument` instead of throwing it here? Otherwise keeping it as is would mirror node.

Reply:
Benefits:
- It is more Java idiomatic
- It removes the indent
- It will probably remove two lines from this method

With that being said, totally optional to change this.

Reply:
changed. thanks for writing it out! I did not realize that checkArgument pattern was a Java idiom.